doc/specs/#885 - Terminal Settings Model.md

```diff
@@ -78,13 +78,126 @@ This separation leaves `AppKeyBindings` with the responsibility of detecting and
  `KeyMapping` handles the (de)serialization and navigation of the key bindings.
 
 
+### Fallback Value
+
+Cascading settings allows our settings model to be constructed in layers (i.e. settings.json values override defaults.json values). With the upcoming introduction of the Settings UI and serialization, it is important to know where a setting value comes from. Consider a Settings UI displaying the following information:
+```json
+    // <profile>: <color scheme value>
+    "defaults": "Solarized", // profiles.defaults
+    "A": "Raspberry", // profile A
+    "B": "Tango", // profile B
+    "C": "Solarized" // profile C
+```
+If `profiles.defaults` gets changed to `"Tango"` via the Settings UI, it is unclear if profile C's value should be updated as well. We need profile C to record if it's value is inherited from profile.defaults or explicitly set by the user.
+
+#### Object Model Inheritance
+
+To start, each settings object will now have a `CreateChild()` function. For `GlobalAppSettings`, it will look something like this:
+```c++
+GlobalAppSettings GlobalAppSettings::CreateChild() const
+{
+    GlobalAppSettings child {};
+    child._parent = this;
+    return child;
+}
+```
+`_parent` serves as a reference for who to ask if a settings value was not provided by the user. `LaunchMode`, for example, will now have a getter/setter that looks similar to this:
+```c++
+// returns the resolved value for this setting
+LaunchMode GlobalAppSettings::LaunchMode()
+{
+    // fallback tree:
+    //  - user set value
+    //  - inherited value
+    //  - system set value
+    return til::coalesce_value(_LaunchMode, _parent.LaunchMode(), LaunchMode::DefaultMode);

```

Comment:
Do we actually need a default value here? Or would there always be one root parent with the default-constructed values that would eventually have the "default" value

Reply:
Probably wouldn't be a root parent -- that would be too complicated, because we'd need the "default constructed" version to have values and the inheriting version to have nullopts, or a different implementation of the tail end of resolution.

This seems the most in line with what we've got without making the CreateChild function set 30 nullopts :smile:

Reply:
This'll be slightly different from what I presented the other day. We won't need a common ancestor that every Profile inherits from. That said, there will still be a root parent for each line. And that root parent will have...
- `std::nullopt` for settings we forgot to define in defaults.json (we fallback to system value automatically)
- the value from defaults.json

Also, the actual implementation isn't using `til::coalesce_value` because we need to check if `parent` exists. But the sentiment is there and reading `til::coalesce_value` is easier to understand than having to read a slightly modified `til::coalesce_value`. So I'm keeping this in there as is. I hope that's ok.